ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/keys/KeyConsistencyValidation.scala

```diff
@@ -65,47 +68,42 @@ private[keys] object KeyConsistencyValidation {
       key: Option[Node.KeyWithMaintainers[Value[ContractId]]],
       targetContractId: Option[ContractId],
       templateId: TypeConName,
-      keyValidationState: KeyValidationState,
-  ): KeyValidationStatus =
+      submittedContractKeysToContractIds: Map[DamlContractKey, Option[RawContractId]],
+  ): Status =
     key match {
-      case None => Right(keyValidationState)
+      case None => Right(submittedContractKeysToContractIds)
       case Some(submittedKeyWithMaintainers) =>
         val submittedDamlContractKey =
           damlContractKey(templateId, submittedKeyWithMaintainers.key)
-        val newKeyValidationState =
-          keyValidationState + ConsistencyKeyValidationState(
-            submittedContractKeysToContractIds = Map(
-              submittedDamlContractKey ->
-                targetContractId.map(_.coid)
+        val newSubmittedContractKeysToContractIds =
+          if (submittedContractKeysToContractIds.contains(submittedDamlContractKey)) {
+            submittedContractKeysToContractIds
+          } else {
+            submittedContractKeysToContractIds.updated(
+              submittedDamlContractKey,
+              targetContractId.map(_.coid),
             )
-          )
+          }
         checkKeyConsistency(
           contractKeysToContractIds,
           submittedDamlContractKey,
-          newKeyValidationState,
+          newSubmittedContractKeysToContractIds,
         )
     }
 
   private def checkKeyConsistency(
       contractKeysToContractIds: Map[DamlContractKey, RawContractId],
       submittedDamlContractKey: DamlContractKey,
-      keyValidationState: KeyValidationState,
-  ): KeyValidationStatus =
+      submittedContractKeysToContractIds: Map[DamlContractKey, Option[RawContractId]],
+  ): Status =
     if (
-      keyValidationState.submittedContractKeysToContractIds
+      submittedContractKeysToContractIds
         .get(
           submittedDamlContractKey
         )
         .flatten != contractKeysToContractIds.get(submittedDamlContractKey)
     )
       Left(Inconsistent)
     else
-      Right(keyValidationState)
-
-  private def ConsistencyKeyValidationState(
-      submittedContractKeysToContractIds: Map[DamlContractKey, Option[
-        RawContractId
-      ]]
-  ): KeyValidationState =
-    KeyValidationState(submittedContractKeysToContractIds)
```

Comment:
Reminder in case you haven’t seen the original issue: This is the bug. We create a KeyValidationState with `activeStateKeys` set to an empty set. `+` always takes `activeStateKeys` from the second argument https://github.com/digital-asset/daml/pull/9941/files#diff-c493ce7c05d6e3f314027a7ab0dad49fec1501c2cfdb0e5b5f861bc916e704c9L139. So this will mean that after each step we reset `activeStateKeys` to an empty set and therefore never get a duplicate.